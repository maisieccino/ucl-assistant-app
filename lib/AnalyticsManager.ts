import * as Amplitude from 'expo-analytics-amplitude'
import Constants from 'expo-constants'

import configureStore from "../configureStore"

const { store } = configureStore

const AMPLITUDE_API_KEY = Constants.manifest.extra
  && Constants.manifest.extra.amplitude
  && Constants.manifest.extra.amplitude.apiKey
  ? Constants.manifest.extra.amplitude.apiKey
  : null

const initialise = (): Promise<void> => Amplitude.initialize(AMPLITUDE_API_KEY)

const shouldTrackAnalytics = (): boolean => (
  AMPLITUDE_API_KEY
  && !__DEV__
  && store.getState().user.settings.shouldTrackAnalytics
)

const setUserId = (userId): void => {
  if (!shouldTrackAnalytics()) { return }

  // make sure this is unique
  if (userId) {
    Amplitude.setUserId(userId)
  }
}

const setUserProperties = ({ ...userProperties }): void => {
  if (!shouldTrackAnalytics()) { return }

  Amplitude.setUserProperties({ ...userProperties })
}

const clearUserProperties = (): void => {
  Amplitude.clearUserProperties()
}

const logEvent = (eventName: string, eventProperties?): void => {
  if (!shouldTrackAnalytics()) { return }

  if (eventProperties) {
    Amplitude.logEventWithProperties(eventName, eventProperties)
  } else {
    Amplitude.logEvent(eventName)
  }
}

const logScreenView = (screenName: string): void => {
  if (!shouldTrackAnalytics()) { return }

  logEvent(`VIEW_SCREEN_${screenName}`)
}

const events = {
  NOTIFICATIONS_ENABLE: `NOTIFICATIONS_ENABLE`,
  NOTIFICATIONS_SKIP: `NOTIFICATIONS_SKIP`,
  PUSH_NOTIFICATIONS_REGISTER: `PUSH_NOTIFICATIONS_REGISTER`,
  SETTINGS_GIVE_FEEDBACK: `SETTINGS_GIVE_FEEDBACK`,
  SETTINGS_RATE_APP: `SETTINGS_RATE_APP`,
  SETTINGS_VIEW_FAQS: `SETTINGS_VIEW_FAQS`,
}

export const AnalyticsManager = {
  clearUserProperties,
  events,
  initialise,
  logEvent,
  logScreenView,
  setUserId,
  setUserProperties,
}

export default AnalyticsManager
