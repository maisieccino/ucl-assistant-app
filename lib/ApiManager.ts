import axios from "axios"
import { encode as btoa } from "base-64"

import {
  EQUIPMENT_URL,
  FREE_ROOMS_URL,
  ROOMBOOKINGS_URL,
  ROOMS_URL,
  SITES_URL,
  TIMETABLE_URL,
  WORKSPACES_URL,
} from "../constants/API"
import ErrorManager from "./ErrorManager"
import LocalisationManager from './LocalisationManager'

const roomsController = {
  getBookings: async (
    token = null,
    {
      roomid,
      siteid,
      date = LocalisationManager.getMoment().format(`YYYYMMDD`),
    },
  ) => {
    if (!roomid || !siteid) {
      throw new Error(`Must specify roomid and siteid`)
    }
    try {
      const results = await axios.get(ROOMBOOKINGS_URL, {
        headers: {
          authorization: `Bearer ${token}`,
        },
        params: {
          date,
          roomid,
          siteid,
        },
      })
      if (!results.data.content.ok) {
        throw new Error(results.data.content.error)
      }
      return results.data.content.bookings
    } catch (error) {
      ErrorManager.captureError(error)
      throw error
    }
  },
  getEmptyRooms: async (
    token = null,
    {
      startDateTime = (new Date()).toISOString(),
      endDateTime = (
        LocalisationManager
          .getMoment()
          .add(1, `hours`)
          .toISOString()
      ),
    } = {},
  ) => {
    try {
      const results = await axios.get(FREE_ROOMS_URL, {
        headers: {
          authorization: `Bearer ${token}`,
        },
        params: {
          endDateTime,
          startDateTime,
        },
      })
      if (!results.data.content.ok) {
        throw new Error(results.data.content.error)
      }
      return results.data.content.free_rooms
    } catch (error) {
      ErrorManager.captureError(error)
      throw error
    }
  },
  getEquipment: async (token = null, { roomid, siteid }) => {
    if (!roomid || !siteid) {
      throw new Error(`Must specify roomid and siteid`)
    }
    try {
      const results = await axios.get(EQUIPMENT_URL, {
        headers: {
          authorization: `Bearer ${token}`,
        },
        params: {
          roomid,
          siteid,
        },
      })
      if (!results.data.content.ok) {
        throw new Error(results.data.content.error)
      }
      return results.data.content.equipment
    } catch (error) {
      ErrorManager.captureError(error)
      throw error
    }
  },
  getSites: async (token = null) => {
    try {
      const results = await axios.get(SITES_URL, {
        headers: {
          authorization: `Bearer ${token}`,
        },
      })
      if (!results.data.content.ok) {
        throw new Error(results.data.content.error)
      }
      return results.data.content.sites
    } catch (error) {
      ErrorManager.captureError(error)
      throw error
    }
  },
  search: async (token = null, query) => {
    if (query && query.length < 3) {
      return {}
    }
    try {
      const results = await axios.get(ROOMS_URL, {
        headers: {
          authorization: `Bearer ${token}`,
        },
        params: {
          query,
        },
      })
      if (!results.data.content.ok) {
        throw new Error(results.data.content.error)
      }
      return results.data.content.rooms
    } catch (error) {
      ErrorManager.captureError(error)
      throw error
    }
  },
}

const workspacesApi = axios.create({
  baseURL: WORKSPACES_URL,
})

const workspacesController = {
  getLiveImage: async (token = null, { surveyId, mapId }) => {
    try {
      const result = await workspacesApi.get(`/getliveimage/map.svg`, {
        headers: {
          authorization: `Bearer ${token}`,
        },
        params: {
          map_id: mapId,
          survey_id: surveyId,
        },
        responseType: `arraybuffer`,
      })
      const image = btoa(
        new Uint8Array(result.data).reduce(
          (data, byte) => data + String.fromCharCode(byte),
          ``,
        ),
      )
      // there may be no need to base64 SVGs
      // but I have yet to test this extensively
      // https://css-tricks.com/probably-dont-base64-svg/
      return `data:image/svg+xml;base64,${image}`
    } catch (error) {
      ErrorManager.captureError(error)
      throw error
    }
  },
  getSummary: async (token = null) => {
    try {
      const { data, headers } = await workspacesApi.get(`/summary`, {
        headers: {
          authorization: `Bearer ${token}`,
        },
      })
      if (!data.content || data.error !== ``) {
        throw new Error(data.error || `There was a problem`)
      }
      return {
        data: data.content,
        lastModified: LocalisationManager.parseToMoment(
          headers[`last-modified`],
        ),
      }
    } catch (error) {
      ErrorManager.captureError(error)
      throw error
    }
  },
  getWorkspaces: async (token = null) => {
    try {
      const { data, headers } = await workspacesApi.get(`/`, {
        headers: {
          authorization: `Bearer ${token}`,
        },
      })
      if (!data.content || data.error !== ``) {
        throw new Error(data.error)
      }
      return {
        data: data.content,
        lastModified: LocalisationManager.parseToMoment(
          headers[`last-modified`],
        ),
      }
    } catch (error) {
      ErrorManager.captureError(error)
      throw error
    }
  },
}

const timetableApi = axios.create({
  baseURL: TIMETABLE_URL,
})

const timetableController = {
  getPersonalTimetable: async (
    token = null,
    date = LocalisationManager.getMoment(),
  ) => {
    const datePart = date
      ? `?date=${date.clone().startOf(`isoWeek`).format(`YYYY-MM-DD`)}`
      : ``
    const headers = {
      authorization: `Bearer ${token}`,
    }
    ErrorManager.addDetail({ headers })
    const results = await timetableApi.get(datePart, { headers })
    if (!results.data.content.ok || results.data.error !== ``) {
      throw new Error(
        results.data.error
        || `There was a problem fetching your personal timetable`,
      )
    }
    const { data: { content: { timetable } } } = results
    Object.keys(timetable).forEach((day) => {
      const dayTimetable = timetable[day]
      timetable[day] = {
        lastModified: LocalisationManager.parseToMoment(
          results.headers[`last-modified`],
        ),
        timetable: dayTimetable,
      }
    })
    return timetable
  },
  getPersonalWeekTimetable: async (
    token = null,
    date = LocalisationManager.getMoment().startOf(`isoWeek`),
  ) => {
    const queryParam = `?date=${date.format(`YYYY-MM-DD`)}`
    const headers = {
      authorization: `Bearer ${token}`,
    }
    ErrorManager.addDetail({ headers })
    const results = await timetableApi.get(`/week/${queryParam}`, {
      headers,
    })
    if (results.data.error !== ``) {
      throw new Error(
        results.data.error
        || `An error occurred when fetching your personal weekly timetable`,
      )
    }
    const { data: { content: { timetable } } } = results
    // eslint-disable-next-line sonarjs/no-identical-functions
    Object.keys(timetable).forEach((day) => {
      const dayTimetable = timetable[day]
      timetable[day] = {
        lastModified: LocalisationManager.parseToMoment(
          results.headers[`last-modified`],
        ),
        timetable: dayTimetable,
      }
    })
    return timetable
  },
}

export default {
  rooms: roomsController,
  timetable: timetableController,
  workspaces: workspacesController,
}
