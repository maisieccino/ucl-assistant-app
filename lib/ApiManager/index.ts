import PeopleController from './PeopleController'
import RoomsController from './RoomsController'
import TimetableController from './TimetableController'
import WorkspacesController from './WorkspacesController'

export default {
  people: PeopleController,
  rooms: RoomsController,
  timetable: TimetableController,
  workspaces: WorkspacesController,
}
