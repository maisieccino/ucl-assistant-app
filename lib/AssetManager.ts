import { Feather } from "@expo/vector-icons"

const uclapi = {
  iconForeground: require(`../assets/images/icon-fg.png`),
  smallIcon: require(`../assets/images/uclapi.png`),
  splash: require(`../assets/images/splash.png`),
}

const undraw = {
  buildingBlocks: require(`../assets/images/undraw_building_blocks.png`),
  calendar: require(`../assets/images/undraw_calendar.png`),
  chilling: require(`../assets/images/undraw_chilling.png`),
  dogWalking: require(`../assets/images/undraw_dog_walking.png`),
  graduation: require(`../assets/images/undraw_graduation.png`),
  notify: require(`../assets/images/undraw_notify.png`),
  peopleSearch: require(`../assets/images/undraw_people_search.png`),
  playfulCat: require(`../assets/images/undraw_playful_cat.png`),
  relaxation: require(`../assets/images/undraw_relaxation.png`),
  relaxingAtHome: require(`../assets/images/undraw_relaxing_at_home.png`),
  scrumBoard: require(`../assets/images/undraw_scrum_board.png`),
  studying: require(`../assets/images/undraw_studying.png`),
}

const font = {
  ...Feather.font,
  apercu: require(`../assets/fonts/somerandomfont.otf`),
  "apercu-bold": require(`../assets/fonts/somerandomfont-Bold.otf`),
  "apercu-light": require(`../assets/fonts/somerandomfont-Light.otf`),
}

export default {
  font,
  uclapi,
  undraw,
}
