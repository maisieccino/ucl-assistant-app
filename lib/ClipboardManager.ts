import { Clipboard } from 'react-native'

const {
  getString,
  setString,
} = Clipboard

export default {
  getString,
  setString,
}
