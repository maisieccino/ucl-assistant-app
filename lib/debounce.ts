// inspiration:
// https://codeburst.io/throttling-and-debouncing-in-javascript-646d076d0a44
const debounceFn = (fn: any, delay: number) => {
  let timer
  return function (...args) {
    if (timer) {
      clearTimeout(timer)
    }
    setTimeout(() => {
      fn(...args)
      timer = null
    }, delay)
  }
}

// from https://github.com/ryanseddon/redux-debounced
// originally released under MIT license
// code copied to prevent Babel errors (unmaintained package)
const redux = () => {
  const timers = {}

  /* eslint-disable consistent-return */
  /* eslint-disable sonarjs/cognitive-complexity */
  const middleware = () => (dispatch) => (action) => {
    const { meta: { debounce = {} } = {}, type } = action

    const {
      time,
      key = type,
      cancel = false,
      leading = false,
      trailing = true,
    } = debounce

    const shouldDebounce = (
      (time && key) || (cancel && key)
    ) && (trailing || leading)
    const dispatchNow = leading && !timers[key]

    const later = (resolve) => () => {
      if (trailing && !dispatchNow) {
        resolve(dispatch(action))
      }
      timers[key] = null
    }

    if (!shouldDebounce) {
      return dispatch(action)
    }

    if (timers[key]) {
      clearTimeout(timers[key])
      timers[key] = null
    }

    if (!cancel) {
      return new Promise((resolve) => {
        if (dispatchNow) {
          resolve(dispatch(action))
        }
        timers[key] = setTimeout(later(resolve), time)
      })
    }
  }
  /* eslint-enable consistent-return */

  /* eslint-disable no-underscore-dangle */
  middleware._timers = timers
  /* eslint-enable no-underscore-dangle */

  return middleware
}

export default { debounce: debounceFn, middleware: redux }
