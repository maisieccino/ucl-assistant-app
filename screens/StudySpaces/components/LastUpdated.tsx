import { Moment } from "moment"
import React from 'react'
import { StyleSheet } from 'react-native'

import { BodyText, ErrorText } from "../../../components/Typography"
import { LocalisationManager } from '../../../lib'

const styles = StyleSheet.create({
  error: {
    marginBottom: 10,
  },
})

interface Props {
  lastModified: Moment | string,
}

class LastUpdated extends React.Component<Props> {
  renderLoading = (): React.ReactElement => <BodyText>Loading...</BodyText>

  renderError = (): React.ReactElement => (
    <ErrorText containerStyle={styles.error}>
      Our studyspace data is stale. Sorry about that.
      We&apos;re working on getting this fixed as quickly as possible.
    </ErrorText>
  )

  render(): React.ReactElement {
    const { lastModified } = this.props
    if (lastModified === null || typeof lastModified !== `object`) {
      return this.renderLoading()
    }

    const isStale = lastModified.isBefore(
      LocalisationManager.getMoment().subtract(5, `minutes`),
    )

    return (
      <>
        {
          isStale ? this.renderError() : null
        }
        <BodyText>
          {`Last updated ${lastModified.calendar().toLowerCase()}`}
        </BodyText>
      </>
    )
  }
}

export default LastUpdated
