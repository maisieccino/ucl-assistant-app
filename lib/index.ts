import AnalyticsManager from './AnalyticsManager'
import ApiManager from './ApiManager'
import AssetManager from './AssetManager'
import debounce from './debounce'
import DeviceManager from './DeviceManager'
import ErrorManager from './ErrorManager'
import LocalisationManager from './LocalisationManager'
import MailManager from './MailManager'
import MapsManager from './MapsManager'
import PushNotificationsManager from './PushNotificationsManager'
import Shadow from './Shadow'
import WebBrowserManager from './WebBrowserManager'

const Random = {
  array: (arr) => arr[Math.floor(Math.random() * arr.length)],
}

export {
  AnalyticsManager,
  ApiManager,
  AssetManager,
  DeviceManager,
  ErrorManager,
  LocalisationManager,
  MailManager,
  MapsManager,
  PushNotificationsManager,
  Random,
  Shadow,
  WebBrowserManager,
  debounce,
}
