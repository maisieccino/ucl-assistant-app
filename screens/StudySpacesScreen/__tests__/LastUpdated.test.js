/**
 * @jest-environment jsdom
 */
import "react-native"

import MockDate from 'mockdate'
import React from 'react'
import { cleanup, render } from "react-native-testing-library"

import { LocalisationManager } from '../../../lib'
import LastUpdated from
  "../components/LastUpdated"

describe(`LastUpdated`, () => {
  MockDate.set(new Date(`2019-11-18T08:47:21`))

  afterAll(() => {
    cleanup()
  })

  it(`renders the LastUpdated component`, () => {
    const mockProps = {
      lastModified: LocalisationManager.getMoment().subtract(4, `minutes`),
    }
    const wrapper = render(<LastUpdated {...mockProps} />)
    expect(wrapper.toJSON()).toMatchSnapshot()
  })

  it(`renders the LastUpdated component loading`, () => {
    const mockProps = {
      lastModified: `${LocalisationManager.parseToDate()}`,
    }
    const wrapper = render(<LastUpdated {...mockProps} />)
    expect(wrapper.toJSON()).toMatchSnapshot()
  })

  it(`shows error when studyspace info is stale`, async () => {
    const mockProps = {
      lastModified: LocalisationManager
        .getMoment()
        .subtract(6, `minutes`),
    }
    const wrapper = render(<LastUpdated {...mockProps} />)
    expect(wrapper.toJSON()).toMatchSnapshot()
  })
})
