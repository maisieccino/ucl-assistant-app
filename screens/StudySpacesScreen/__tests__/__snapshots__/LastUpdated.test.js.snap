// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LastUpdated renders the LastUpdated component 1`] = `
<Text
  style={
    Array [
      Object {
        "color": "#4D5061",
        "fontFamily": "apercu",
        "fontSize": 16,
        "marginBottom": 1,
        "marginTop": 1,
      },
      undefined,
    ]
  }
>
  Last updated today at 8:43 am
</Text>
`;

exports[`LastUpdated renders the LastUpdated component loading 1`] = `
<Text
  style={
    Array [
      Object {
        "color": "#4D5061",
        "fontFamily": "apercu",
        "fontSize": 16,
        "marginBottom": 1,
        "marginTop": 1,
      },
      undefined,
    ]
  }
>
  Loading...
</Text>
`;

exports[`LastUpdated shows error when studyspace info is stale 1`] = `
Array [
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "center",
        },
        Object {
          "alignContent": "center",
          "alignItems": "center",
          "flexDirection": "row",
          "marginBottom": 10,
          "marginTop": 5,
        },
      ]
    }
  >
    <Text />
    <Text
      style={
        Array [
          Object {
            "color": "rgba(215, 38, 61, 1)",
            "flex": 1,
            "fontFamily": "apercu",
            "fontSize": 16,
            "marginBottom": 1,
            "marginLeft": 5,
            "marginTop": -1,
          },
          undefined,
        ]
      }
    >
      Our studyspace data is stale. Sorry about that. We're working on getting this fixed as quickly as possible.
    </Text>
  </View>,
  <Text
    style={
      Array [
        Object {
          "color": "#4D5061",
          "fontFamily": "apercu",
          "fontSize": 16,
          "marginBottom": 1,
          "marginTop": 1,
        },
        undefined,
      ]
    }
  >
    Last updated today at 8:41 am
  </Text>,
]
`;
